### ISSUE TITLE
pcm overflows for ogg files

### ISSUE THREAD

**Issue by derselbst** [erikd/libsndfile#194]
I came across an unexpected behaviour when decoding loud ogg files. If a sample seems to have max value, it will overflow when decoding it with sndfile's integer read functions. The image shows the waveform of an [example ogg file](https://raw.githubusercontent.com/SuperTux/data/1bbcf668304466b37868356e0d0278f2132c90d2/music/fortress.ogg) directly imported to audacity. The waveform right below shows the same ogg file decoded to raw `.wav` before import.
![sndfile_bug](https://cloud.githubusercontent.com/assets/8152480/22437015/9c072c12-e726-11e6-9f2e-6b37fcc800bb.png)

To reproduce:
* grab the [example file](https://raw.githubusercontent.com/SuperTux/data/1bbcf668304466b37868356e0d0278f2132c90d2/music/fortress.ogg)
* play it back using `sndfile-play`
* on linux, everything should sound fine so far, since `sndfile-play` uses `sf_read_float()` here
* adjust `sndfile-play` to make it decoding using  `sf_read_int()` (also occurs with `sf_read_short()`)
* when listening again you should hear the first crack after about 8 seconds

To avoid the overflow I have to run `SFC_SET_SCALE_FLOAT_INT_READ`, however (from my understanding) [documentation](http://www.mega-nerd.com/libsndfile/api.html#note2) only suggests this when having `SF_FORMAT_FLOAT` or `SF_FORMAT_DOUBLE`.

So if it is intended to run this command also for `SF_FORMAT_VORBIS`, an update of docs might be fine? Or am I missing something?

**Comment by erikd** [comment]
Yes, I think the documentation should be updated. Will keep this ticket open until I do that.

**Comment by kcat** [comment]
Having the option to clamp the integer samples instead of wrapping would be good, too. The distortion from clamped samples is far less noticeable and less disturbing compared to wrapped samples, but having to attenuate the whole file to avoid a few out of range samples seems like overkill.

Also something else I noticed relating to pcm overflows, as mentioned in kcat/openal-soft#505, reading ogg samples as int scales like `pcm [n][j] * 2147483647.0f`. But single-precision floats can't represent 2147483647, so it gets rounded at compile-time to the closest representable value, `2147483648.0f`. When +1.0 is scaled and converted to int, 2147483648 wraps to -2147483648, causing a file that even stays within the -1...+1 range to overflow sample values on int reads.

**Comment by derselbst** [comment]
> Having the option to clamp the integer samples instead of wrapping would be good, too.

No, we don't need this, because it would only hide misusage of libsndfile's API even more. What we _do_ need is an update of the documentation. Even 4 years later I still have no clue when exactly I have to call `SFC_SET_SCALE_FLOAT_INT_READ`.

@erikd pls. come back to this.

**Comment by evpobr** [comment]
@derselbst , pull request?

**Comment by kcat** [comment]
> No, we don't need this, because it would only hide misusage of libsndfile's API even more.

Attenuating the whole (potentially minutes-long) file to avoid an occasional transient sample wrapping isn't always ideal, especially if it requires decoding and scanning through the whole file first before decoding the samples again to actually give to the caller. This leaves me with having to decode to float and convert to integer myself just in case, which makes the library that much less nice/efficient to use.

**Comment by derselbst** [comment]
> Attenuating the whole (potentially minutes-long) file to avoid an occasional transient sample wrapping isn't always ideal, especially if it requires decoding and scanning through the whole file first before decoding the samples again to actually give to the caller

I agree that is not ideal. But this is the way how libsndfile works. Which is why I think that your use-case is beyond the scope of this issue.

I just want to know in which cases I have to execute `SFC_SET_SCALE_FLOAT_INT_READ`. But I don't know because OGG files report as:

```gdb
print/x sfinfo.format
$2 = 0x200060
```

Which is `SF_FORMAT_VORBIS | SF_FORMAT_OGG`.

 I would expect it to report as 
`SF_FORMAT_VORBIS | SF_FORMAT_OGG | SF_FORMAT_FLOAT`. Then it would be clear that it contains floats and has to be normalized via `SFC_SET_SCALE_FLOAT_INT_READ`.

> derselbst , pull request?

I really have no time to go through the internals of  libsndfile to find all those codecs that require `SFC_SET_SCALE_FLOAT_INT_READ`. But I'm sure Erik could tell within a few mintues.

**Comment by kcat** [comment]
> I just want to know in which cases I have to execute `SFC_SET_SCALE_FLOAT_INT_READ`. But I don't know because OGG files report as:
> 
> ```gdb
> print/x sfinfo.format
> $2 = 0x200060
> ```
> 
> Which is `SF_FORMAT_VORBIS | SF_FORMAT_OGG`.
> 
> I would expect it to report as
> `SF_FORMAT_VORBIS | SF_FORMAT_OGG | SF_FORMAT_FLOAT`. Then it would be clear that it contains floats and has to be normalized via `SFC_SET_SCALE_FLOAT_INT_READ`.

The problem there is both `SF_FORMAT_VORBIS` and `SF_FORMAT_FLOAT` are within the format sub-type mask. Different sub-type enumerations can share bits, so it couldn't be extended safely, to e.g. `SF_FORMAT_OPUS` as that uses the same bit as `SF_FORMAT_PCM_32` despite Opus natively decoding to floating-point. The only safe/future-proof way I can see it working is to have a query for a format's native decode type, its decoded bit-depth and whether it's normalized integer or floating-point (you could just add checks in your code for `SF_FORMAT_VORBIS`, `SF_FORMAT_OPUS`, etc, along with `SF_FORMAT_FLOAT` and `SF_FORMAT_DOUBLE`, but as new formats get added that use floating-point, like MP3, old code would have to be updated along with the library for glitch-free support).
